sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -132,13 +132,13 @@ case class InSubqueryExec(
   def updateResult(): Unit = {
     val rows = plan.executeCollect()
     result = child.dataType match {
-      case _: StructType => rows.toArray
-      case _ => rows.map(_.get(0, child.dataType))
+      case _: StructType => rows.toSet
```

Comment:
This doesn't match the `InSet` implementation. For non-atomic type, we need to build a `TreeSet` to recursively compare the elements of struct/array/map.

Reply:
Yes, fix it by https://github.com/apache/spark/pull/29475/commits/d1db0bc231eb87a09699a8d3f99e3e64ce2885c6